packages/web/server/queries/complex/assets/price.ts

```diff
@@ -236,3 +241,86 @@ export async function calcAssetValue({
 }
+
+const tokenHistoricalPriceCache = new LRUCache<string, CacheEntry>(
+  DEFAULT_LRU_OPTIONS
+);
+/** Cached query function for getting an asset's historical price for a given token and time frame. */
+export async function getAssetHistoricalPrice({
+  coinDenom,
+  timeFrameMinutes,
+  numRecentFrames,
+}: {
+  /** Major (symbol) denom to fetch historical price data for. */
+  coinDenom: string;
+  /** Number of minutes per bar. So 60 refers to price every hour. */
+  timeFrameMinutes: TimeFrame;
+  /** How many recent price values to splice with.
+   *  For example, with `timeFrameMinutes` set to every hour (60) and `numRecentFrames` set to 24, you get the last day's worth of hourly prices. */
+  numRecentFrames?: number;
+}): Promise<TokenHistoricalPrice[]> {
+  return await cachified({
+    cache: tokenHistoricalPriceCache,
+    key: `token-historical-price-${coinDenom}-${timeFrameMinutes}-${
+      numRecentFrames ?? "all"
+    }`,
+    ttl: 1000 * 60 * 3, // 3 minutes
+    getFreshValue: () =>
+      numRecentFrames
+        ? queryTokenHistoricalChart({
+            coinDenom,
+            timeFrameMinutes: timeFrameMinutes,
+          }).then((prices) => prices.slice(-numRecentFrames))
+        : queryTokenHistoricalChart({
+            coinDenom,
+            timeFrameMinutes: timeFrameMinutes,
+          }),
+  });
+}
+
+export type CommonHistoricalPriceTimeFrame = "1H" | "1D" | "1W" | "1M";
+/** Get's recent historical price data given common time frame literals.
+ *  The configurations for each time frame are as follows:
+ *
+ *  - "1H": 5-minute bars, last hour of prices (12 recent frames)
+ *  - "1D": 1-hour bars, last day of prices (24 recent frames)
+ *  - "1W": 12-hour bars, last week of prices (14 recent frames)
+ *  - "1M": 1-day bars, last month of prices (30 recent frames)
+ */
+export async function getCommonTimeFrameAssetHistoricalPrice({
+  coinDenom,
+  timeFrame,
+}: {
+  coinDenom: string;
+  timeFrame: CommonHistoricalPriceTimeFrame;
+}) {
+  let timeFrameMinutes;
+  switch (timeFrame) {
+    case "1H":
+      timeFrameMinutes = 5 as TimeFrame; // 5 minute bars
+    case "1D":
+      timeFrameMinutes = 60 as TimeFrame; // 1 hour bars
+    case "1W":
+      timeFrameMinutes = 720 as TimeFrame; // 12 hour bars
+    case "1M":
+      timeFrameMinutes = 1440 as TimeFrame; // 1 day bars
+  }
+  timeFrameMinutes = timeFrameMinutes as TimeFrame;
+
+  let numRecentFrames;
+  if (timeFrame === "1H") {
+    numRecentFrames = 12; // Last hour of prices in 5 bars of minutes
+  } else if (timeFrame === "1D") {
+    numRecentFrames = 24; // Last day of prices with bars of 60 minutes
+  } else if (timeFrame === "1W") {
+    numRecentFrames = 14; // Last week of prices with bars of 12 hours
+  } else if (timeFrame === "1M") {
+    numRecentFrames = 30; // Last month of prices with bars as 1 day
+  }
+
+  return getAssetHistoricalPrice({
+    coinDenom,
+    timeFrameMinutes,
+    numRecentFrames,
+  });
+}
```

Comment:
What are your thoughts on merging this function with `getAssetHistoricalPrice` and handle these timeframes in a conditional? Handling nested functions can be quite taxing to understand.